crates/codegen/spec/src/productions.rs

```diff
@@ -189,6 +205,23 @@ fn write_subexpression<T: Write>(
     }
 }
 
+fn get_precedence(expression: &ExpressionRef) -> u8 {
```

Comment:
F2f: add an extension trait, and rename `get_precedence` to `precedence`

Reply:
OR rename to expression_precedence, especially given it's a private function. just a nit.